### PR TITLE
Source NAT option on Private Gateway

### LINKED ISSUE #2680: Using Source NAT option on Private Gateway does not work
##### ISSUE TYPE

 * Bug Report

##### COMPONENT NAME

~~~
VPC (Private Gateway Source NAT)
~~~

##### CLOUDSTACK VERSION

~~~
At least: 4.9, 4.10, and 4.11
~~~

##### CONFIGURATION

When you use the Source NAT feature of Private Gateways on a VPC.  This should Source NAT all traffic from CloudStack VMs going towards IPs reachable through Private Gateways

##### OS / ENVIRONMENT

I happen to be running KVM, but shouldn't be hypervisor specific.  it is an issue in the SystemVM / VPC Router

##### SUMMARY

There is a bug in the Private Gateway functionality, when Source NAT is enabled for the Private Gateway.  When the SNAT is added to iptables, it has the source CIDR of the private gateway subnet.  Since no VMs live in that private gateway subnet, the SNAT doesn’t work.  

##### STEPS TO REPRODUCE

Below is an example:

- VMs have IP addresses in the 10.0.0.0/24 subnet.
- The Private Gateway address is 10.101.141.2/30
 
In the outputs below, the SOURCE field for the new SNAT (eth3) only matches if the source is 10.101.141.0/30.  Since the VM has an IP address in 10.0.0.0/24, the VMs don’t get SNAT’d as they should when talking across the private gateway.  The SOURCE should be set to ANYWHERE.
##### BEFORE ADDING PRIVATE GATEWAY
~~~
Chain POSTROUTING (policy ACCEPT 1 packets, 52 bytes)
pkts bytes target     prot opt in     out     source               destination
    2   736 SNAT       all  --  any    eth2    10.0.0.0/24          anywhere             to:10.0.0.1
   16  1039 SNAT       all  --  any    eth1    anywhere             anywhere             to:46.99.52.18
~~~

##### EXPECTED RESULTS

~~~
Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
pkts bytes target     prot opt in     out     source               destination
    0     0 SNAT       all  --  any    eth3    anywhere             anywhere             to:10.101.141.2
    2   736 SNAT       all  --  any    eth2    anywhere             anywhere             to:10.0.0.1
   23  1515 SNAT       all  --  any    eth1    anywhere             anywhere             to:46.99.52.18
~~~

##### ACTUAL RESULTS

~~~
Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
pkts bytes target     prot opt in     out     source               destination
    0     0 SNAT       all  --  any    eth3    10.101.141.0/30      anywhere             to:10.101.141.2
    2   736 SNAT       all  --  any    eth2    10.0.0.0/24          anywhere             to:10.0.0.1
   23  1515 SNAT       all  --  any    eth1    anywhere             anywhere             to:46.99.52.18
~~~
##### SUGGESTED FIX
It looks like CsAddress.py treats the creation of the Private Gateway SNAT as if it is a GUEST network, which works fine, except for the SNAT problem shown above.  Here is the code from MASTER (line 479 is SNAT rule):
~~~
if self.get_type() in ["guest"]:
...
...
    self.fw.append(["nat", "front",
        "-A POSTROUTING -s %s -o %s -j SNAT --to-source %s" %
        (guestNetworkCidr, self.dev, self.address['public_ip'])])
 ~~~

I am thinking we just change that to the following.  I can’t think of any reason we need the source/guest CIDR specified:
~~~
if self.get_type() in ["guest"]:
...
...
    self.fw.append(["nat", "front",
        "-A POSTROUTING -o %s -j SNAT --to-source %s" %
        (self.dev, self.address['public_ip'])])
~~~

##### More Information
Here is a link I found that further documents the Private Gateway Source NAT feature:

https://cwiki.apache.org/confluence/display/CLOUDSTACK/Source+NAT+on+private+gateway

Notice, how in the SNAT shown it does not specify a source guest network CIDR - my above fix matches the example in the link above.

### PR DESCRIPTION
**Pull request by Slair1**
Using Source NAT option on Private Gateway does not work
This fixes #2680 

## Description

When you use the Source NAT feature of Private Gateways on a VPC.  This should Source NAT all traffic from CloudStack VMs going towards IPs reachable through Private Gateways.

This change in this PR, stops adding the Source CIDR to SNAT rules.  This should be discussed/reviewed, but i can see no reason why the Source CIDR is needed.  There can only be one SNAT IP per interface, except for Static (one-to-one) NATs, which still work with this change in place.  The outbound interface is what matters in the rule.

##### SUMMARY

There is a bug in the Private Gateway functionality, when Source NAT is enabled for the Private Gateway.  When the SNAT is added to iptables, it has the source CIDR of the private gateway subnet.  Since no VMs live in that private gateway subnet, the SNAT doesn’t work.  

##### STEPS TO REPRODUCE

Below is an example:

- VMs have IP addresses in the 10.0.0.0/24 subnet.
- The Private Gateway address is 10.101.141.2/30
 
In the outputs below, the SOURCE field for the new SNAT (eth3) only matches if the source is 10.101.141.0/30.  Since the VM has an IP address in 10.0.0.0/24, the VMs don’t get SNAT’d as they should when talking across the private gateway.  The SOURCE should be set to ANYWHERE.
##### BEFORE ADDING PRIVATE GATEWAY
~~~
Chain POSTROUTING (policy ACCEPT 1 packets, 52 bytes)
pkts bytes target     prot opt in     out     source               destination
    2   736 SNAT       all  --  any    eth2    10.0.0.0/24          anywhere             to:10.0.0.1
   16  1039 SNAT       all  --  any    eth1    anywhere             anywhere             to:46.99.52.18
~~~

##### EXPECTED RESULTS

~~~
Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
pkts bytes target     prot opt in     out     source               destination
    0     0 SNAT       all  --  any    eth3    anywhere             anywhere             to:10.101.141.2
    2   736 SNAT       all  --  any    eth2    anywhere             anywhere             to:10.0.0.1
   23  1515 SNAT       all  --  any    eth1    anywhere             anywhere             to:46.99.52.18
~~~

##### ACTUAL RESULTS

~~~
Chain POSTROUTING (policy ACCEPT 0 packets, 0 bytes)
pkts bytes target     prot opt in     out     source               destination
    0     0 SNAT       all  --  any    eth3    10.101.141.0/30      anywhere             to:10.101.141.2
    2   736 SNAT       all  --  any    eth2    10.0.0.0/24          anywhere             to:10.0.0.1
   23  1515 SNAT       all  --  any    eth1    anywhere             anywhere             to:46.99.52.18
~~~
## Types of changes

- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] New feature (non-breaking change which adds functionality)
- [X] Bug fix (non-breaking change which fixes an issue)
- [ ] Enhancement (improves an existing feature and functionality)
- [ ] Cleanup (Code refactoring and cleanup, that may add test cases)

## GitHub Issue/PRs

Fixes: #2680 

## Checklist:

- [x] I have read the [CONTRIBUTING](https://github.com/apache/cloudstack/blob/master/CONTRIBUTING.md) document.
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
Testing
- [ ] I have added tests to cover my changes.
- [ ] All relevant new and existing integration tests have passed.
- [ ] A full integration testsuite with all test that can run on my environment has passed.

### CODE REVIEW

**Comment by ustcweizhou** [comment]
@Slair1 
CsAddress.py has been moved to systemvm/debian/opt/cloud/bin/cs/CsAddress.py in master

**Comment by rhtyd** [comment]
@Slair1 can you rebase this against 4.11 branch? /cc @DaanHoogland @PaulAngus @borisstoyanov   - this may explain some of the private gateway ACL related failures, we'll need to check if those smoketest failures are valid errors or env caused issues.

**Comment by rhtyd** [comment]
ping @Slair1 can you rebase your branch?

**Comment by PaulAngus** [comment]
ping @Slair1 can you rebase your branch please if you'd like this included in 4.11.1, otherwise we'll have to push it out to 4.12/4.11.2

**Comment by Slair1** [comment]
@PaulAngus I've rebased to 4.11

**Comment by rhtyd** [comment]
@blueorangutan package

File: systemvm/debian/opt/cloud/bin/cs/CsAddress.py [hunk 1]
```diff
@@ -477,8 +477,8 @@ def fw_vpcrouter(self):
                             ("eth1", "eth1", guestNetworkCidr)])
 
             self.fw.append(["nat", "front",
-                            "-A POSTROUTING -s %s -o %s -j SNAT --to-source %s" %
-                            (guestNetworkCidr, self.dev, self.address['public_ip'])])
+                            "-A POSTROUTING -o %s -j SNAT --to-source %s" %
+                            (self.dev, self.address['public_ip'])])
```
Comment (rhtyd) [inline systemvm/debian/opt/cloud/bin/cs/CsAddress.py]: @Slair1 to minimize impact/regression from this, can you consider @ustcweizhou 's advise on doing something like : 
```
if self.address["source_nat"]:
    self.fw.append(new rule)
else:
    self.fw.append(old rule)
```

**Comment by rhtyd** [comment]
@blueorangutan test

File: systemvm/debian/opt/cloud/bin/cs/CsAddress.py (hunk 1 above)
Comment (rafaelweingartner) [reply systemvm/debian/opt/cloud/bin/cs/CsAddress.py]: I think it might be safer to use @rhtyd suggestion. What do you guys think about it?

Comment (GabrielBrascher) [reply systemvm/debian/opt/cloud/bin/cs/CsAddress.py]: I agree with you guys. @Slair1 can you please address @rhtyd comment?

**Comment by rhtyd** [comment]
@Slair1 can you test if the change I made per Wei's comment works in your env. I'll rekick tests.
@blueorangutan package

File: systemvm/debian/opt/cloud/bin/cs/CsAddress.py [hunk 2]
```diff
@@ -476,7 +476,12 @@ def fw_vpcrouter(self):
             self.fw.append(["", "front", "-A NETWORK_STATS_%s -o %s -s %s" %
                             ("eth1", "eth1", guestNetworkCidr)])
 
-            self.fw.append(["nat", "front",
+            if self.address["source_nat"]:
```
Comment (rhtyd) [inline systemvm/debian/opt/cloud/bin/cs/CsAddress.py]: @ustcweizhou @rafaelweingartner @GabrielBrascher can you help review and test this? /cc @PaulAngus @DaanHoogland

**Comment by PaulAngus** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
@blueorangutan package

**Comment by DaanHoogland** [comment]
@blueorangutan test

File: systemvm/debian/opt/cloud/bin/cs/CsAddress.py (hunk 2 above)
Comment (Slair1) [reply systemvm/debian/opt/cloud/bin/cs/CsAddress.py]: @rhtyd See my reply on the Issue #2680

**Comment by PaulAngus** [comment]
tried manual tests to see why failed so badly.
when trying to created a VPC received the following error:

image

**Comment by ustcweizhou** [comment]
@PaulAngus because a ")" is missing at the end of line 482 ......

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
The test_03_vpc_privategw_restart_vpc_cleanup is the oly failure that seems related to this PR, however the output only says "Internal Server Error", so no real conclusions to take from that. Restarting to see persistence of the issue
@blueorangutan test

**Comment by Slair1** [comment]
squashed all the commits into 1

**Comment by rhtyd** [comment]
@blueorangutan package

**Comment by rhtyd** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
The test_03_vpc_privategw_restart_vpc_cleanup is the only failure that seems related to this PR, however the output only says "Internal Server Error", so no real conclusions to take from that. Restarting to see persistence of the issue
@blueorangutan test

**Comment by DaanHoogland** [comment]
:( the error is consistent. need to look

File: systemvm/debian/opt/cloud/bin/cs/CsAddress.py [hunk 3]
```diff
@@ -476,9 +476,10 @@ def fw_vpcrouter(self):
             self.fw.append(["", "front", "-A NETWORK_STATS_%s -o %s -s %s" %
                             ("eth1", "eth1", guestNetworkCidr)])
 
-            self.fw.append(["nat", "front",
-                            "-A POSTROUTING -s %s -o %s -j SNAT --to-source %s" %
-                            (guestNetworkCidr, self.dev, self.address['public_ip'])])
+            if self.address["source_nat"]:
+                self.fw.append(["nat", "front",
```
Comment (rhtyd) [inline systemvm/debian/opt/cloud/bin/cs/CsAddress.py]: @Slair1 please fix indent:
```
../debian/opt/cloud/bin/cs/CsAddress.py:481:29: E128 continuation line under-indented for visual indent
../debian/opt/cloud/bin/cs/CsAddress.py:482:29: E128 continuation line under-indented for visual indent
+'[' 1 -gt 0 ']'
+echo 'Pylint failed, please check your code'
Pylint failed, please check your code
```

Comment (Slair1) [reply systemvm/debian/opt/cloud/bin/cs/CsAddress.py]: @rhtyd Increased indent and squashed commits into 1

**Comment by DaanHoogland** [comment]
@blueorangutan package

File: systemvm/debian/opt/cloud/bin/cs/CsAddress.py (hunk 3 above)
Comment (Slair1) [reply systemvm/debian/opt/cloud/bin/cs/CsAddress.py]: @DaanHoogland It looks like the test "test_03_vpc_privategw_restart_vpc_cleanup" is failing on several PRs.  Looking at the logs it seems to be a NullPointerException:

```
java.lang.NullPointerException
	at com.cloud.network.router.NicProfileHelperImpl.createPrivateNicProfileForGateway(NicProfileHelperImpl.java:95)
```

### NicProfileHelperImpl.java (Lines 93 - 95)
```
            final PrivateIpAddress ip =
                    new PrivateIpAddress(ipVO, privateNetwork.getBroadcastUri().toString(), privateNetwork.getGateway(), netmask,
                            NetUtils.long2Mac(NetUtils.createSequenceBasedMacAddress(ipVO.getMacAddress(), NetworkModel.MACIdentifier.value())));
```

**Comment by DaanHoogland** [comment]
@blueorangutan test

**Comment by DaanHoogland** [comment]
both these two errors are known. The private gateway one is related though.

**Comment by DaanHoogland** [comment]
@rhtyd why did you repackage this? It wasn't changed since the last successful package run.

**Comment by PaulAngus** [comment]
@DaanHoogland - as we know the cause of the test_privategw_acl failures - are we ok with this PR?

**Pull request merged**